File: frontend/src/lib/api/billing.ts
import { apiClient } from "@/lib/api/client";

export type PlanTier = "free" | "pro";

export interface Subscription {
  plan: PlanTier;
  status: string;
  is_pro: boolean;
  current_period_end: string | null;
}

export interface BillingConfig {
  configured: boolean;
  publishable_key: string | null;
}

export const billingApi = {
  config: async (): Promise<BillingConfig> =>
    (await apiClient.get<BillingConfig>("/billing/config")).data,
  subscription: async (): Promise<Subscription> =>
    (await apiClient.get<Subscription>("/billing/subscription")).data,
  checkout: async (): Promise<string> =>
    (await apiClient.post<{ url: string }>("/billing/checkout")).data.url,
  portal: async (): Promise<string> =>
    (await apiClient.post<{ url: string }>("/billing/portal")).data.url,
  sync: async (): Promise<Subscription> =>
    (await apiClient.post<Subscription>("/billing/sync")).data,
};

File: backend/app/schemas/billing.py
"""Billing / subscription schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.subscription import PlanTier


class SubscriptionPublic(BaseModel):
    """The company's current plan and billing status."""

    model_config = ConfigDict(from_attributes=True)

    plan: PlanTier
    status: str
    is_pro: bool
    current_period_end: datetime | None


class BillingConfig(BaseModel):
    """Whether billing is enabled on this deployment (for the frontend)."""

    configured: bool
    publishable_key: str | None = None


class CheckoutResponse(BaseModel):
    """A Stripe-hosted URL to redirect the browser to."""

    url: str
